### ISSUE TITLE
import errors not handled?

### ISSUE THREAD

**Issue by nbecker** [pybind/pybind11#656]
This demonstrates an import error gives core dump.  I think it should instead give a python ImportError, as python normally would:
```
>>> import blah
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named 'blah'
```

bug2.cc:
```
#include "pybind11/numpy.h"     // vectorize
#include "pybind11/pybind11.h"
#include "numpy/arrayobject.h"
//#include "ndarray/pybind11.h"
#include <complex>

//namespace nd=ndarray;
namespace py = pybind11;


PYBIND11_PLUGIN (bug2) {
  if (_import_array() < 0) {
    PyErr_SetString(PyExc_ImportError, "numpy.core.multiarray failed to import");
    return nullptr;
  }

  py::module m("bug1", "pybind11 example plugin");

  py::module::import ("blah");

  return m.ptr();
}
```
```
 python3 -c 'import bug2'
terminate called after throwing an instance of 'pybind11::error_already_set'
  what():  SystemError: <built-in method __contains__ of dict object at 0x7fded2d14f48> returned a result with an error set
Aborted (core dumped)
```

**Comment by dean0x7d** [comment]
Thanks for the example code. I believe PR #657 should fix it. You can try pulling from that branch and see if it works. It should yield a proper error message now:
```
ImportError: ModuleNotFoundError: No module named 'blah'
```

**Comment by nbecker** [comment]
Sorry, I'm a git nube.  What do I need to do to test this patch ("pull from that branch")?  According to github help, looking here https://github.com/pybind/pybind11/pull/657 I should see a link for "command line" that should answer my question, but it's not there.

**Comment by dean0x7d** [comment]
If you don't have an existing pybind11 repository, you'll need to:
```
git clone -b <branch> <remote_repo>
```
where you want the relevant branch on my fork:
```
git clone -b fix-import-error https://github.com/dean0x7d/pybind11.git
```

It's a bit different if you have an existing pybind11 repository:
```
git remote add -f <name> <url>
git checkout <branch>
```
where you want:
```
git remote add -f dean0x7d https://github.com/dean0x7d/pybind11.git
git checkout fix-import-error
```

**Comment by nbecker** [comment]
Yes, my problem is fixed with this patch.
BTW, my use case is,
equalizer has classes that inherit from fir3 classes.  So equalizer.cc uses
py::import("blah.fir3")
Is this the preferred approach?

**Comment by nbecker** [comment]
I also see that [https://github.com/pybind/pybind11/pull/657](url) includes patches for noconvert, so I tried it on my code.  It doesn't work for my use case.
```
 m.def("mag_sqr", [](py::array_t<double> a) {
      print ("double");
      return py::vectorize([](double x) {return x * x;})(a);
    },
    py::arg("in").noconvert()
    );
  m.def("mag_sqr", [](py::array_t<float> a) {
      print ("float");
      return py::vectorize([](float x) {return x * x;})(a);
    },
    py::arg("in").noconvert()
    );
  m.def("mag_sqr", [](py::array_t<std::complex<double>> a) {
      print ("complex");
      return py::vectorize([](std::complex<double> x) {return real(x) * real(x) + imag(x) * imag(x);})(a);
    },
    py::arg("in").noconvert()
    );
  m.def("mag_sqr", [](py::array_t<std::complex<float>> a) {
      return py::vectorize([](std::complex<float> x) {return real(x) * real(x) + imag(x) * imag(x);})(a);
    },
    py::arg("in").noconvert()
    );
 
```
In my test:
```
u = np.arange (12, dtype=float).reshape (3,4)
v = mag_sqr (u.real)
w = mag_sqr (u.astype(complex))
x = mag_sqr (u.astype(np.float32))
```

I get:
```
double
/home/nbecker/.local/bin/ipython3:7: ComplexWarning: Casting complex values to real discards the imaginary part
  from IPython import start_ipython
double
double
```

**Comment by dean0x7d** [comment]
The `array_t` part hasn't been merged yet. See #650. You can try out those changes and report any issues there.

**Comment by jagerman** [comment]
> where you want the relevant branch on my fork: ...

A very useful alternative that I use is to run:

`git config --add remote.upstream.fetch "+refs/pull/*/head:refs/remotes/upstream/pr/*"`

replacing the two `upstream`s with whatever the main pybind/pybind11 remote is named in your local checkout.  Then a `git fetch upstream` will fetch all PRs, and allow you to use `git checkout pr/657` to checkout a PR without needing to add the PR author's repository as another remote.

**Comment by nbecker** [comment]
@dean0x7d I did
```
 git remote add -f jagerman https://github.com/jagerman/pybind11.git
 git merge jagerman/array_t-overload-resolution
error: Terminal is dumb, but EDITOR unset
Not committing merge; use 'git commit' to complete the merge.
[nbecker@nbecker2 pybind11]$ git commit
error: Terminal is dumb, but EDITOR unset
Please supply the message using either -m or -F option.
[nbecker@nbecker2 pybind11]$ git commit -m merge
[fix-import-error d13f015] merge
```
Now test_mag_sqr.py says:
```
python3 test_mag_sqr.py 
double
complex
float
```
which seems to be working correctly, and it seems my import error from #657 is also fixed.

**Comment by dean0x7d** [comment]
> @jagerman said
> A very useful alternative that I use is to run: ...

Cool, I didn't know that. That's much more convenient. Thanks!
